lib/upgrade/etcd/conversionv1v3/ippool.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package conversionv1v3
+
+import (
+	"fmt"
+	"strings"
+
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/ipip"
+	cnet "github.com/projectcalico/libcalico-go/lib/net"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func ConvertIPPool(kvp *model.KVPair) (*model.KVPair, error) {
+	pool, ok := kvp.Value.(model.IPPool)
+	if !ok {
+		return nil, fmt.Errorf("value is not a valid BGPPeer resource key")
+	}
+
+	return &model.KVPair{
+		Key: model.ResourceKey{
+			Name: cidrToName(pool.CIDR),
+			Kind: strings.ToLower(apiv3.KindIPPool),
+		},
+		Value: apiv3.IPPool{
+			ObjectMeta: v1.ObjectMeta{Name: cidrToName(pool.CIDR)},
+			Spec: apiv3.IPPoolSpec{
+				CIDR:        pool.CIDR.String(),
+				IPIPMode:    convertIPIPMode(pool.IPIPMode),
+				NATOutgoing: pool.Masquerade,
+				Disabled:    pool.Disabled,
+			},
+		},
+	}, nil
+}
+
+func convertIPIPMode(mode ipip.Mode) apiv3.IPIPMode {
+	switch strings.ToLower(string(mode)) {
+	case "", "off":
+		return apiv3.IPIPModeNever
+	case "cross-subnet":
+		return apiv3.IPIPModeCrossSubnet
+	case "always":
```

Comment:
Hmm, just looking at the v2 code and I think there is another bug in the Get processing.  The IPIPMode was added at a later stage so that means it's possible to have a backend model that has the Tunnel interface set (i.e. IPIP is enabled), but that there is no IPIPMode field -- in which case we should default to "Always".

Therefore I think "tunl" set and no IPIPMode also corresponds to "Always".

The lack of a "tunl" interface names is what corresponds to "Never"